extensions/mongodb/src/main/java/com/hazelcast/jet/mongodb/WriteMongoP.java

```diff
@@ -0,0 +1,627 @@
+        }
+        this.replaceOptions = options;
+        this.connectionRetryTracker = new RetryTracker(RETRY_STRATEGY);
+
+        collectionPicker = new ConstantCollectionPicker(databaseName, collectionName);
+    }
+
+    /**
+     * Creates a new processor that will choose database and collection for each item.
+     */
+    public WriteMongoP(
+            SupplierEx<? extends MongoClient> connectionSupplier,
+            FunctionEx<I, String> databaseNameSelectFn,
+            FunctionEx<I, String> collectionNameSelectFn,
+            Class<I> documentType,
+            FunctionEx<I, Object> documentIdentityFn,
+            ConsumerEx<ReplaceOptions> replaceOptionAdjuster,
+            String documentIdentityFieldName
+    ) {
+        this.documentType = documentType;
+        this.connectionSupplier = connectionSupplier;
+        this.documentIdentityFn = documentIdentityFn;
+        this.documentIdentityFieldName = documentIdentityFieldName;
+        ReplaceOptions options = new ReplaceOptions().upsert(true);
+        if (replaceOptionAdjuster != null) {
+            replaceOptionAdjuster.accept(options);
+        }
+        this.replaceOptions = options;
+        this.connectionRetryTracker = new RetryTracker(RETRY_STRATEGY);
+
+        collectionPicker = new FunctionalCollectionPicker(databaseNameSelectFn, collectionNameSelectFn);
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) {
+        logger = context.logger();
+
+        int processorIndex = context.globalProcessorIndex();
+        transactionUtility = new UnboundedTransactionsProcessorUtility<>(
+                getOutbox(),
+                context,
+                context.processingGuarantee(),
+                () -> new MongoTransactionId(context, processorIndex),
+                txId -> new MongoTransaction(txId, logger),
+                txId -> {
+                    MongoTransaction mongoTransaction = activeTransactions.get(txId);
+                    refreshTransaction(mongoTransaction, true);
+                },
+                () -> {
+                    for (MongoTransaction tx : activeTransactions.values()) {
+                        refreshTransaction(tx, false);
+                    }
+                }
+        );
+    }
+
+    @Override
+    public boolean tryProcessWatermark(@Nonnull Watermark watermark) {
+        return true;
+    }
+
+    @Override
+    protected boolean tryProcess(int ordinal, @Nonnull Object item) {
+        return transactionUtility.tryProcess();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void process(int ordinal, @Nonnull Inbox inbox) {
+        if (!reconnectIfNecessary()) {
+            return;
+        }
+        MongoTransaction mongoTransaction = transactionUtility.activeTransaction();
+
+        ArrayList<I> items = new ArrayList<>();
+        for (Object item : inbox) {
+            items.add((I) item);
+            if (items.size() >= MAX_BATCH_SIZE) {
```

Comment:
What happens with items beyond MAX_BATCH_SIZE?

Reply:
Will be processed in the next invocation of `complete()`